### DISCUSSION TITLE
Bootcode for running tests on non-QEMU models

### DISCUSSION THREAD

**Discussion by ostannard** [picolibc/picolibc#850] [Q&A]
Hi Keith,

As part of https://github.com/ARM-software/LLVM-embedded-toolchain-for-Arm, we'd like to be able to run the picolibc tests on non-QEMU models, because we want to support some Arm architecture versions which don't have full QEMU support. In particular, I'm trying to use Arm's free FVP (fixed virtual platform) models, so that our tests can be run by contributors outside of Arm.

The current picolibc crt0 code seems to work for M-profile FVPs, with just a modified linker script to match the memory map. However, the AArch64 crt0 contains code to set up a page table, which makes assumptions about the memory map which aren't correct for that FVP. The RAM and device regions are in the opposite order to QEMU, so it's not just a case of adjusting the addresses. Another target I'd like to get working (but haven't tried yet) is R-profile AArch64, which isn't supported by QEMU, and doesn't currently have any compatible bootcode in picolibc.

It would also be useful for us to provide an easier way for users to provide their own bootcode. The code in `picocrt/crt0.h` is target-independent, and should work for any models and hardware. However, the code in `picocrt/machine` will need to be changed to run code on different hardware. The former is an incline function included in the latter, so re-using it requires a copy of the picolibc source code.

The options for solving this I've thought of so far are:

1. Add new variants of the crt0 code in picolibc, with different target-specific code. I don't expect this to be an acceptable answer, since it's not done for any other targets, and wouldn't be useful to users.
2. Split the crt0 into two objects, one containing the target-independent code from crt0.h, and one containing the hardware-specific parts. We could then build one crt0 library containing both, for testing with QEMU, and another with just the target-independent parts. We'd then add options to provide a bootcode object which can be used for testing on any real or emulated hardware, and build our own bootcode outside of picolibc for the FVP models.
3. Make the existing bootcode more general, so that it gets all memory addresses from the linker script. The only change I know of so far is that the AArch64 code would need to build the page table at runtime, based on the actual location of the RAM. I've not tried doing this yet, so it has the risk that we end up discovering lots of other differences between QEMU and FVPs, complicating the startup code.

What do you think of these options? We'd like to keep our downstream patches on picolibc as minimal as possible, so I'd like to find a solution to this which works well for you.

Thanks,
Oliver

**Comment by keith-packard** [comment]
Getting more test coverage would be really useful, so I'm happy to help make that happen. Frankly, what ever makes things easiest for you, both for initial development and ongoing maintenance. The crt0 code in picolibc performs two key functions:

 1. Enable testing via emulation
 2. Provide sample code that developers can start with

Many "real" picolibc projects will likely need their own startup code, so expecting that picolibc will provide code they can use unchanged is not a strong requirement. I'd rather the code were easy to understand than having it super optimized for production use.

Given the options you've outlined, I think either options 1 or 3 would make the most sense. Option 3 looks like it might end up with less duplication of code? And if that doesn't work out in the long run, we can always use option 1 as needed.

I'd do whatever was most convenient for you right now so that we can get experience using FVP; if we don't like how it looks, we can fix it later. Especially if we find a way to run FVP in picolibc CI; if that's possible, then we'll do what's necessary to make it work.

**Comment by ostannard** [comment]
Thanks, after implementing this for AArch64 I now think option 1 is going to be easiest, because the FVP models boot up at a different exception level to QEMU, so different versions of every system register need to be used. That's easier to do with macros to build the appropriate version for the target, than switching all of those system register accesses at runtime. I've pushed what I've got so far for review at #852.
